aws-distro-opentelemetry-node-autoinstrumentation/src/llo-handler.ts

```diff
@@ -0,0 +1,558 @@
+    if (!span.events) {
+      return;
+    }
+
+    const updatedEvents: TimedEvent[] = [];
+
+    for (const event of span.events) {
+      if (!event.attributes) {
+        updatedEvents.push(event);
+        continue;
+      }
+
+      const updatedEventAttributes = this.filterAttributes(event.attributes);
+
+      if (Object.keys(updatedEventAttributes).length !== Object.keys(event.attributes).length) {
+        const updatedEvent: TimedEvent = {
+          name: event.name,
+          time: event.time,
+          attributes: updatedEventAttributes,
+        };
+
+        updatedEvents.push(updatedEvent);
+      } else {
+        updatedEvents.push(event);
+      }
+    }
+
+    (span as any).events = updatedEvents;
+  }
+
+  /**
+   * Extract LLO attributes and emit them as a single consolidated Gen AI Event.
+   *
+   * This method:
+   * 1. Collects all LLO attributes using the pattern registry
+   * 2. Groups them into input and output messages
+   * 3. Emits one event per span containing all LLO content
+   *
+   * The event body format:
+   * {
+   *   "input": {
+   *     "messages": [
+   *       {
+   *         "role": "system",
+   *         "content": "..."
+   *       },
+   *       {
+   *         "role": "user",
+   *         "content": "..."
+   *       }
+   *     ]
+   *   },
+   *   "output": {
+   *     "messages": [
+   *       {
+   *         "role": "assistant",
+   *         "content": "..."
+   *       }
+   *     ]
+   *   }
+   * }
+   *
+   * @param span The source ReadableSpan containing the attributes
+   * @param attributes LLO attributes to process
+   * @param eventTimestamp Optional timestamp to override span timestamps
+   * @returns
+   */
+  private emitLloAttributes(span: ReadableSpan, attributes: Attributes, eventTimestamp?: HrTime): void {
+    if (!attributes || Object.keys(attributes).length === 0) {
+      return;
+    }
+
+    //[] maybe add sanity check here from Python "has_llo_attrs"
+
+    const allMessages = this.collectAllLloMessages(span, attributes);
+    if (allMessages.length === 0) {
+      return;
+    }
+
+    // Group messages into input/output categories
+    const groupedMessages = this.groupMessagesByType(allMessages);
+
+    // Build event body
+    const eventBody: AnyValue = {};
+    if (groupedMessages.input.length > 0) {
+      eventBody.input = { messages: groupedMessages.input };
+    }
+    if (groupedMessages.output.length > 0) {
+      eventBody.output = { messages: groupedMessages.output };
+    }
+
+    if (Object.keys(eventBody).length === 0) {
+      return;
+    }
+
+    // Create and emit the event
+    const timestamp = eventTimestamp || span.endTime;
+    const eventLogger = this.eventLoggerProvider.getEventLogger(span.instrumentationLibrary.name);
+
+    // Hack - Workaround to add a custom-made Context to an Event so that the emitted event log
+    // has the correct associated traceId and spanId. This is needed because a ReadableSpan only
```

Comment:
does it also include sampling flag?

Reply:
Updated comment to also mention including `flags`.